text/0003-performance-data.md

```diff
@@ -0,0 +1,151 @@
+If the profiler flags are enabled the following will need to happen: 
+Transports will need to output the number of bytes being sent in the current tick and the latency. The data will be written out to a ProfilerTickData chunk. This data will be generic and transport agnostic.
+MLAPI will need to output the number of networked objects and the amount of RPCs being sent. This data will be added to the ProfilerTickData chunk.
+
+External Tools and APIs will be able to then get this fully written ProfilerTickData at the end of the frame for them to do what they wish (Render to Profiler Stats Overlay or Display/Record in a Profiler or file)
+
+As this data is largely a Dictionary, accessing its data will be as simple as accessing things by specified fields. Additionally the rendering or logging of these statistics will fall on the user and not affect the Tick or be done at all within the MLAPI framework.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+
+In order to limit the interface, we are pushing the collected set or performance data to be obtainable only at the end of a tick. The following is roughly what this object could look like. Currently this object will be updated in ```NetworkManager.LateUpdate()``` for the MLAPI and ```Transport.Send()``` and other relevant transport method for the transport layer. At the end of the tick the user will be notified that the relevant data is complete and ready (both the MLAPI data and Transport data). Since MLAPI is responsible for collecting the data, and the data from transports will be generic and apply to all/most transports, MLAPI as  the owner of the collection of performance data will call ```GetTransportGetData()``` on the transports through an interface in order to get the data exactly when it needs it in the pipeline before propagating the data out externally. Whether or not the data is collected per transport can also be toggled by a runtime flag ‘profilerEnabled’.
+
+```cs
+namespace MLAPI {
+	class ProfilerConstants {
+		public const string NumberOfServerRPCs = "numberOfServerRPCs";
+
+		public const string TickDuration = "tickDuration";
+		public const string NumberOfBuffMessagesOut = "numberOfBuffMessagesOut";
+		public const string NumberOfBuffMessagesIn = "numberOfBuffMessagesIn";
+		public const string NumberOfUnBuffMessagesOut = "numberOfUnBuffMessagesOut";
+		public const string NumberOfUnBuffMessagesIn = "numberOfUnBuffMessagesIn";
+		public const string NumberBytesSent = "numberBytesSent";
+        public const string NumberBytesReceived = "numberBytesReceived";
+	}
+
+	public class ProfilerTickData {
+		public int tickID;
+
+		public readonly Dictionary<string, int> tickData = new Dictionary<string, int>();
+	}
+
+	public interface ITransportProfilerData {
+		Dictionary<string, int> GetTransportGetData();
```

Comment:
This is an implementation detail and IMO doesn't block this RFC, but we should pass in the dictionary rather than having it return it so that we can re-use the dictionary between network ticks and avoid memory allocs

Reply:
I see below that the unet transport example does cache its own dictionary. this approach could work as well, although I think this adds complexity because the internal dictionary is used externally and the transport doesn't know when it is safe to clear the dictionary for the next network tick

Reply:
Yes the transport is in control of its own memory block and deciding what goes into it. It will cache regardless.
And yes i'm willing to refactor here in an actual PR here for performance gains. Passing in a dictionary to the transport might be useful in terms of clarity even though it will still be caching locally.

Reply:
I'd like to add a similar implementation-detail question here — it might be too constraining to map `string` to `int` but not `string` to `object`. Since performance won't be a big concern here (throwing boxing cost out of the window), it might be useful (and more future-proof) to map `string` to `object` and allow passing other types such as `float` etc through this proposed design.

Reply:
I only somewhat agree. But for something that is typically a float (like timestamp) we can just represent with milliseconds. And outside of that i failed to come up with an example that required floats or a vector (which you can just split it up into three ints) 

Having a contrarian example (which i don't have) would help me take a stronger position one way or the other

Reply:
I think I'm still 51% on `string`—`int` mapping side but wanted to call out just in case. I'll try to convince both myself and you as well if I could find better examples.